app/controllers/users_controller.rb

```diff
@@ -42,13 +43,24 @@ def destroy
     redirect_to users_url, notice: notice_message
   end
 
+  def edit_password; end
+
+  def update_password
+    if @user.update(user_params)
+      sign_in(@user, bypass: true)
+      redirect_to edit_password_users_url, notice: 'Password was successfully updated.'
+    else
+      render :edit_password
+    end
+  end
+
   private
 
   def user_params
+    allowed_params = [:email, :password]
+    allowed_params << :role if current_user.admin?
     params.require(:user).permit(
-      :email,
-      :password,
-      :role
+      allowed_params
```

Comment:
If we use a net new controller for this, we dont need to worry about the conditional modification of the params allowlist.